--- /root/aurora-1.8.8/main.c	2015-01-27 02:37:17.397517079 +0100
+++ /root/my-aurora-1.8.8/main.c	2015-08-18 14:07:38.885727145 +0200
@@ -414,18 +414,18 @@
             if (rPID == PID) fprintf (stderr, " = me");
             fprintf (stderr, "\n");
         }
-        if (command != NULL) {
-            free(command);
-            command = NULL;
-        }
         if (rPID != PID) {
-             if (SubStrPos == NULL) {
+             if (command == NULL) {	// Clear stale only rPID process is dead 
                  if (bVerbose) fprintf (stderr, "\n");
                  fprintf(stderr, "%s: %s: Clearing stale serial port lock. (%lu)\n",getCurTime(),ProgramName,rPID);
                  ClrSerLock(rPID);
              } else if (yLockWait > 0)
                  sleep(1);
         }
+        if (command != NULL) {
+            free(command);
+            command = NULL;
+        }
     }
     if (bVerbose && rPID == PID) fprintf(stderr, "Appears we got the lock.\n");
     if (rPID != PID) {
